### ISSUE TITLE
Can you release the training code?

### ISSUE THREAD

**Issue by qhykwsw** [guohengkai/region-ensemble-network#24]
Thank you for your awesome research. I want to use your method to train my own model but I find that there is only the prediction code in the repo. Can you release the training code together? I will be very grateful for that.

**Comment by qhykwsw** [comment]
Hi, sorry to bother again. I am a newcomer for caffe and I have a few questions. First, When training the data, did you only add the new data and loss layer for the project to original caffe, anything else? Second, did you put the data augmentation in the data layer you edited by yourself? Is there any way else can achieve the same goal in caffe?

**Comment by guohengkai** [comment]
@qhykwsw Hi, sorry for the late response.

> First, When training the data, did you only add the new data and loss layer for the project to original caffe, anything else? 

We added a new data layer, a loss layer and add more augmentation functions in the data transformer.

> Second, did you put the data augmentation in the data layer you edited by yourself? 

We put them into the data transformer.

**Comment by qhykwsw** [comment]
I see, thanks for your advices. Can you share the new data layer file to me? My email address  is qianhongyi@buaa.edu.cn. If you cannot, does it need to write .cu file for the new data layer?

**Comment by guohengkai** [comment]
@qhykwsw Hi, the data layer contains lots of other things that are used in my lab, so I'm afraid that I could not share it with you now. Sorry for that. It is only for CPU, not .cu.

**Comment by qhykwsw** [comment]
Nevermind, I guess I can take this opportunity to learn caffe. Thanks again and wish you do better research.

**Comment by guohengkai** [comment]
@qhykwsw Thanks. I think the results are not difficult to reproduce. Welcome to ask any questions if necessary.

**Comment by qhykwsw** [comment]
Thanks for your patience. In fact, I do have two question. First, I am curious about when do the data augmentation? Is that after the cube was cropped or directly on the orign picture? Second, if you do the data augmentation after the cube was cropped, when doing the translation within [−10, 10] pixel, the joints may become out of the cube. What should I do about this situation?

**Comment by guohengkai** [comment]
First, we did it after cropping to reduce the times of transformation.

Second, I think it didn't harm the network too much when the joints are slightly out of the cube. And you can also set the cube large enough to avoid this situation.

**Comment by qhykwsw** [comment]
Ok, I see. So, what's the best augmentation order? First translation and then scaling and then rotation? Or it's really doesn't matter? And when doing the scaling and rotation, what's the reference point, center of the picture, or the center of mass calculated by the previous step?

**Comment by guohengkai** [comment]
It doesn't matter. You can try it:) The center is the one of cropping.

**Comment by qhykwsw** [comment]
Ok, last question. When doing the augmentation, for example translation, there will be some pixels newly go into the cube, how should I set these pixels, 0, 255 or any value else?

**Comment by guohengkai** [comment]
It depends. We set them to background values.

**Comment by qhykwsw** [comment]
I see and I will try to reproduce the research follow your setting, thank you so much.

**Comment by dhecloud** [comment]
Hi, I am also trying to augment my data. Just curious, after translating the depth image, eg to the right by 10 pixels, do you also change the ground truth joint (x,y,z) coordinate to reflect that? 

if yes, how do you do it? do you just add +10?

**Comment by guohengkai** [comment]
@dhecloud Sure, the joint should also be changed by add 10. Otherwise the results are wrong.

**Comment by qhykwsw** [comment]
Hi, I noticed that the dropout layer in your model(deploy.prototxt) didn't specify the phase. Does that mean the model do dropout in both training and testing phase?

**Comment by guohengkai** [comment]
@qhykwsw The layer itself will take care of it. You can view the layer codes for details.

**Comment by qhykwsw** [comment]
I did not expect this feature. I specify the dropout layer just work in the trainning phase as usual.
